### DISCUSSION TITLE
Model data not updated on second save()

### DISCUSSION THREAD

**Discussion by svdigital-development** [phalcon/cphalcon#15625] [General]
We are upgrading our application from Phalcon 3.4.5 to Phalcon 4.1.2 and from PHP 7.3 to PHP 7.4.

Everything work fine except for one serious problem.

We have a model "`Request`" which uses `protected` properties and `public` getters and setters and with Exception on `save()` failure. There is a Controller which updates the model twice on an already existing record, using another method of the Controller class inside a transaction.

Code:

```php
$db = $this->di->get( 'db' ); // using ::get() or ::getShared() doesn't change; the DB connection service is shared (we use ::setShared() on the DI)

$db->begin();

try {
	/* Just returns an instance of RequestStatus model (existing record, of course) */
	$newStatus = self::calculateNewStatus( $request );

	/* FIRST SAVE(): SUCCESSFULL */
	$this->evaluateApply( $request );
	
	/* Then, we assign a NEW value to "RequestStatusId" */
	$request->setRequestStatusId( $newStatus->getId() );
	
	/* SECON SAVE(): RETURNS TRUE BUT "RequestStatusId" IS NOT UPDATED! */
	$request->save();

	$db->commit();
	
} catch ( \Throwable $ex ) {
	$db->rollback();
	throw $ex;
}

// ...

private function evaluateApply( Request $request ) {
	$request->setUserAssignedId( $this->getCurrentUserId() );
	$request->save();
}
```

As you can see the `$request` record is updated (`save`d) twice: once inside another method which accepts the model instance as parameter and then inside the main method.

This code has always worked great.

**PROBLEM**

After the second `$request->save()` the instance of `$request` contains:

- The NEW `UserAssignedId` (first save)
- The OLD `RequestStatusId` (second save)

Dirty state is alway "TRANSIENT" on all stages.

Please note that we tried `var_dump()` the model data before and after every statement. The `RequestStatusId` value actually changes. It is correct until the beginning of the `save()` method (we extended the `save()` method just to `var_dump()` the value of the field before calling `return parent::save()`).

The value is the NEW one until before `return parent::save()`. Then the value changes back to the old one already inside the first model events: `beforeValidation()`, `afterValidationOnUpdate()`, `beforeSave()`, `afterSave()`.

The only way to make the model perform the correct second update is to do this:

```php
	/* See below for this method */
	$this->evaluateApply( $request );
	
	$request->refresh();
	
	$request->setRequestStatusId( $newStatus->getId() );
```

I mean, we must `refresh()` the model before performing the subsequent assignment. This is obviously unacceptable because we have plenty of situations of this kind in our application. And it is unacceptable because we're working with objects...

Please, can you help us?

Thank you

**Comment by Jeckerson** [comment]
Hello, could you also provide example your setters: `setRequestStatusId()` and `setUserAssignedId()`?

**Reply by Jeckerson** [reply]
With mine example I can't simulate your problem:
```php
// Controller

protected function update($model)
{
    $model->setColumn1('first');
    $model->save();
}

public function indexAction()
{
    $this->db->begin();
    $model = TableModel::findFirst(1);
    $this->update($model);

    $model->setColumn2('second');
    $model->save();
    $this->db->commit();
}

```

```php
// Model

protected ?string $column1 = null;
protected ?string $column2 = null;

public function setColumn1($val)
{
    $this->column1 = $val;
}

public function setColumn2($val)
{
    $this->column2 = $val;
}

```

**Reply by svdigital-development** [reply]
Hello,
yes but they are very simple. And please note that the above code is unchanged since years, I mean it is working fine on production environment with php 7.3 and Phalcon 3.

```php
public function setRequestStatusId( int $statusId ): Request {
	$this->RequestStatusId = $statusId;
	
	return $this;
}


public function setUserAssignedId( int $userId ): Request {
	$this->UserAssignedId= $userId;
	
	return $this;
}
```

Thank you

**Reply by svdigital-development** [reply]
Ok thanks. I see you can't reproduce the issue, but why is this happening just upgrading to Phalcon 4 without any change to our code?

**Reply by Jeckerson** [reply]
Could you contact me in our Discord? - https://phalcon.io/discord

**Reply by svdigital-development** [reply]
It seems like Phalcon ORM is using the previous "version" of the object when performing the second `save()`, just when it enters the "Zephir" method `save()` (https://github.com/phalcon/cphalcon/blob/master/phalcon/Mvc/Model.zep#L2456).

I extended the `save()` method inside the Model class, just to print data for debug purposes and then returning the parent's `save()`.

Results of the `var_dump()` just before `return parent::save()`, inside the extended `save()` in the Model:

- The value of "UserAssignedId" is the NEW one (just saved before)
- The value of "RequestStatusId" is the NEW one (both using property or getter method)
- There is ONE "changedField()": "RequestStatusId"
- The snapshot value of "RequestStatusId" is the previous one

Making the same `var_dump()`s at the beginning of `beforeValidation()` or `beforeValidationOnUpdate()` inside the Model class tell us that:

- The value of "UserAssignedId" is the NEW one (just saved before)
- The value of "RequestStatusId" is the OLD one (both using property or getter method)
- There is NO "changedField()"s
- The snapshot value of "RequestStatusId" is the previous one

**Comment by svdigital-development** [comment]
I found what **causes** the problem. However, I still did not understand why it happens and how to solve it. 

`Request` model contains this code inside `initialize()`:


```php
$this->belongsTo( 'RequestStatusId', RequestStatus::class, 'Id', [
	'alias' => 'RequestStatus',
	'reusable' => false,
] );
```

Let's see the code in the first message of the thread (https://github.com/phalcon/cphalcon/discussions/15625#discussion-3530621).

That code is 100% working, except for the first method invocation:

```php
	/* Just returns an instance of RequestStatus model (existing record, of course) */
	$newStatus = self::calculateNewStatus( $request );
```

Actually, `self::calculateNewStatus( $request )` is doing this:

```php
return StatusSet::calculateStatus( $request->getRequestStatus()->getNewStatusCode() );
```

The point is that **doing this**:

```php
$request->getRequestStatus();
```

**just before doing this** (see the code in the first message):

```php
$request->setRequestStatusId( $newStatus->getId() );
$request->save();
```

Causes the problem that `$request->RequestStatusId;` is still the old one.

In the code above I can reproduce the problem **(1)** or I can correctly save the record **(2)**:

**(1)**:

```php
var_dump( $request->getRequestStatusId() ); // returns "111"

/* Make the ORM perform the foreign key query */
$request->getRequestStatus();

/* This method saves $request (another field, not "RequestStatusId"): successfull */
$this->evaluateApply( $request );

$request->setRequestStatusId( 222 );

$request->save();

var_dump( $request->getRequestStatusId() ); // returns "111"
```

**(2)**:

```php
var_dump( $request->getRequestStatusId() ); // returns "111"

/* DON'T perform the foreign key query */
/*** $request->getRequestStatus; ***/

/* This method saves $request (another field, not "RequestStatusId"): successfull */
$this->evaluateApply( $request );

$request->setRequestStatusId( 222 );

$request->save();

var_dump( $request->getRequestStatusId() ); // returns "222"
```

**Reply by wurst-hans** [reply]
I didn't look into this forum for a long time, but I've faced exactly the same problem long ago. Great, that you've filed an issue that is going to be resolved in Phalcon v5. But I'm still using Phalcon v4, because upgrading would be much effort for now. So, if anyone stumbled across this problem, here is, how I solved this in v4.

Instead of assigning relation the way you did it using `$request->setRequestStatusId($newStatus->getId());` (which is the preferred method for me too), I had to assign whole entity to virtual relation property, like `$request->RequestStatus = $newStatus;`. This will successfully store the new relation.

**Comment by svdigital-development** [comment]
I created an issue for this: https://github.com/phalcon/cphalcon/issues/15649

**Comment by svdigital-development** [comment]
### To reproduce

https://github.com/phalcon/cphalcon/issues/15649#issuecomment-915864151
